File: backend/app/api/admin_estate.py
"""Super-admin estate endpoints: summary dashboard and audit trail."""

from __future__ import annotations

import csv
import io
import json
from datetime import UTC, datetime, time
from datetime import date as date_t
from typing import Annotated, Any, Literal
from zoneinfo import ZoneInfo

from fastapi import APIRouter, Query
from fastapi.responses import Response
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.config import settings
from app.deps import SessionDep, SuperAdminUser
from app.errors import Forbidden
from app.models.audit import AuditLog
from app.models.checklist import InspectionEntry
from app.models.entry import Entry
from app.models.user import User
from app.schemas.admin_estate import (
    AdminSummaryOut,
    AuditLogList,
    AuditLogOut,
    EstateTotals,
    SegmentOut,
    SegmentTotals,
    SiteSummaryOut,
)
from app.services import admin_summary, audit

router = APIRouter(prefix="/admin", tags=["admin"])

PeriodMode = Literal["today", "week", "month", "custom"]
IST = ZoneInfo(settings.timezone)

REGISTER_LABELS = {
    "work_done": "Work Done",
    "coolant": "Coolant",
    "driver_complaint": "Driver Complaint",
    "breakdown": "Breakdown",
    "pm_schedule": "PM Schedule",
}


def _parse_day(raw: str | None) -> date_t | None:
    if not raw:
        return None
    return date_t.fromisoformat(raw)


def _day_bounds_ist(day: date_t, *, end: bool) -> datetime:
    """UTC-aware instant at start (or end) of an IST calendar day."""
    wall = time(23, 59, 59, 999999) if end else time(0, 0, 0)
    local = datetime.combine(day, wall, tzinfo=IST)
    return local.astimezone(UTC)


@router.get("/summary", response_model=AdminSummaryOut)
async def estate_summary(
    _user: SuperAdminUser,
    session: SessionDep,
    period: Annotated[PeriodMode, Query()] = "month",
    date_from: Annotated[str | None, Query()] = None,
    date_to: Annotated[str | None, Query()] = None,
    month: Annotated[str | None, Query()] = None,
) -> AdminSummaryOut:
    """Site-wise incident counts for the Admin Summary and project reports."""
    start, end = admin_summary.resolve_period(
        period,
        date_from=_parse_day(date_from),
        date_to=_parse_day(date_to),
        month=month,
    )
    raw = await admin_summary.build_summary(session, date_from=start, date_to=end)
    await session.commit()

    def _site(s: dict) -> SiteSummaryOut:
        return SiteSummaryOut(**s)

    segments = {
        key: SegmentOut(
            label=seg["label"],
            totals=SegmentTotals(**seg["totals"]),
            sites=[_site(s) for s in seg["sites"]],
        )
        for key, seg in raw["segments"].items()
    }
    return AdminSummaryOut(
        date_from=raw["date_from"],
        date_to=raw["date_to"],
        estate=EstateTotals(**raw["estate"]),
        sites=[_site(s) for s in raw["sites"]],
        segments=segments,
    )


def _subject_from_after(object_type: str, after: dict[str, Any] | None) -> str | None:
    if not after:
        return None
    if object_type == "entry":
        register = after.get("register")
        label = REGISTER_LABELS.get(str(register), str(register) if register else "Entry")
        bus = after.get("bus_no") or ""
        site = after.get("site") or ""
        day = after.get("date") or ""
        parts = [p for p in (label, bus, site, day) if p]
        return " · ".join(parts) if parts else None
    if object_type == "inspection":
        code = after.get("work_type") or after.get("work_type_name") or "Inspection"
        bus = after.get("bus_no") or after.get("vehicle") or ""
        site = after.get("site") or ""
        day = after.get("inspected_on") or ""
        km = after.get("milestone_km")
        km_s = f"{km // 1000}k" if isinstance(km, int) and km >= 1000 else (
            str(km) if km is not None else ""
        )
        parts = [p for p in (str(code), bus, site, day, km_s) if p]
        return " · ".join(parts) if parts else None
    if object_type == "fitted_unit":
        unit = after.get("unit") or "Unit"
        bus = after.get("vehicle") or ""
        return " · ".join(p for p in (f"Fitted {unit}", bus) if p)
    return None


def _audit_out(
    row: AuditLog,
    actors: dict[str, User],
    *,
    entry_by_id: dict[str, Entry] | None = None,
    inspection_by_id: dict[str, InspectionEntry] | None = None,
) -> AuditLogOut:
    actor = actors.get(row.actor_id) if row.actor_id else None
    before = json.loads(row.before) if row.before else None
    after = json.loads(row.after) if row.after else None
    if not isinstance(before, dict):
        before = None
    if not isinstance(after, dict):
        after = None

    register: str | None = None
    site_code: str | None = None
    bus_no: str | None = None
    subject: str | None = None

    if after:
        register = after.get("register") if isinstance(after.get("register"), str) else None
        site_code = after.get("site") if isinstance(after.get("site"), str) else None
        bus_no = after.get("bus_no") or after.get("vehicle")
        if isinstance(bus_no, str) and not bus_no:
            bus_no = None
        subject = _subject_from_after(row.object_type, after)

    # Hydrate older rows that only stored form fields (no register/site).
    if row.object_type == "entry" and entry_by_id:
        entry = entry_by_id.get(row.object_id)
        if entry is not None:
            register = register or entry.register.value
            site_code = site_code or entry.site_code
            bus_no = bus_no or (
                entry.vehicle.registration_no if entry.vehicle else None
            )
            if not subject:
                label = REGISTER_LABELS.get(entry.register.value, entry.register.value)
                subject = " · ".join(
                    p
                    for p in (
                        label,
                        bus_no or "",
                        entry.site_code,
                        entry.entry_date.isoformat(),
                    )
                    if p
                )
            # Merge into after so CSV/export also carries register.
            after = {
                **(after or {}),
                "register": register,
                "site": site_code,
                "date": entry.entry_date.isoformat(),
                "bus_no": bus_no,
                "status": entry.status.value,
            }

    if row.object_type == "inspection" and inspection_by_id:
        insp = inspection_by_id.get(row.object_id)
        if insp is not None:
            site_code = site_code or insp.site_code
            bus_no = bus_no or (
                insp.vehicle.registration_no if insp.vehicle else None
            )
            code = insp.work_type.code if insp.work_type else "Inspection"
            if not subject:
                subject = " · ".join(
                    p
                    for p in (
                        code,
                        bus_no or "",
                        insp.site_code,
                        insp.inspected_on.isoformat(),
                    )
                    if p
                )
            after = {
                **(after or {}),
                "site": site_code,
                "bus_no": bus_no,
                "vehicle": bus_no,
                "work_type": code,
                "inspected_on": insp.inspected_on.isoformat(),
                "milestone_km": insp.milestone_km,
            }

    return AuditLogOut(
        id=row.id,
        actor_id=row.actor_id,
        actor_user_id=actor.user_id if actor else None,
        actor_name=actor.name if actor else None,
        action=row.action,
        object_type=row.object_type,
        object_id=row.object_id,
        before=before,
        after=after,
        created_at=row.created_at,
        subject=subject,
        register=register,
        site_code=site_code,
        bus_no=bus_no,
    )


async def _load_actors(session: SessionDep, rows: list[AuditLog]) -> dict[str, User]:
    ids = {r.actor_id for r in rows if r.actor_id}
    if not ids:
        return {}
    users = (await session.scalars(select(User).where(User.id.in_(ids)))).all()
    return {u.id: u for u in users}


async def _load_entries(
    session: SessionDep, rows: list[AuditLog]
) -> dict[str, Entry]:
    ids = {r.object_id for r in rows if r.object_type == "entry"}
    if not ids:
        return {}
    found = (
        await session.scalars(
            select(Entry)
            .where(Entry.id.in_(ids))
            .options(selectinload(Entry.vehicle))
        )
    ).all()
    return {e.id: e for e in found}


async def _load_inspections(
    session: SessionDep, rows: list[AuditLog]
) -> dict[str, InspectionEntry]:
    ids = {r.object_id for r in rows if r.object_type == "inspection"}
    if not ids:
        return {}
    found = (
        await session.scalars(
            select(InspectionEntry)
            .where(InspectionEntry.id.in_(ids))
            .options(
                selectinload(InspectionEntry.vehicle),
                selectinload(InspectionEntry.work_type),
            )
        )
    ).all()
    return {i.id: i for i in found}


@router.get("/audit", response_model=AuditLogList)
async def list_audit(
    user: SuperAdminUser,
    session: SessionDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=200)] = 50,
    actor_id: Annotated[str | None, Query()] = None,
    action: Annotated[str | None, Query()] = None,
    object_type: Annotated[str | None, Query()] = None,
    date_from: Annotated[str | None, Query()] = None,
    date_to: Annotated[str | None, Query()] = None,
) -> AuditLogList:
    """User-wise audit trail — every recorded action, newest first."""
    if not user.has_permission("em_audit:read"):
        raise Forbidden("Missing permission: em_audit:read")

    start = _parse_day(date_from)
    end = _parse_day(date_to)
    rows, total = await audit.list_logs(
        session,
        actor_id=actor_id,
        action=action,
        object_type=object_type,
        date_from=_day_bounds_ist(start, end=False) if start else None,
        date_to=_day_bounds_ist(end, end=True) if end else None,
        offset=(page - 1) * page_size,
        limit=page_size,
    )
    actors = await _load_actors(session, rows)
    entries = await _load_entries(session, rows)
    inspections = await _load_inspections(session, rows)
    await session.commit()
    return AuditLogList(
        items=[
            _audit_out(r, actors, entry_by_id=entries, inspection_by_id=inspections)
            for r in rows
        ],
        page=page,
        page_size=page_size,
        total=total,
    )


@router.get("/audit/export")
async def export_audit(
    user: SuperAdminUser,
    session: SessionDep,
    actor_id: Annotated[str | None, Query()] = None,
    action: Annotated[str | None, Query()] = None,
    object_type: Annotated[str | None, Query()] = None,
    date_from: Annotated[str | None, Query()] = None,
    date_to: Annotated[str | None, Query()] = None,
) -> Response:
    """CSV of the filtered audit trail (capped)."""
    if not user.has_permission("em_audit:read"):
        raise Forbidden("Missing permission: em_audit:read")

    start = _parse_day(date_from)
    end = _parse_day(date_to)
    rows, _ = await audit.list_logs(
        session,
        actor_id=actor_id,
        action=action,
        object_type=object_type,
        date_from=_day_bounds_ist(start, end=False) if start else None,
        date_to=_day_bounds_ist(end, end=True) if end else None,
        offset=0,
        limit=5_000,
    )
    actors = await _load_actors(session, rows)
    entries = await _load_entries(session, rows)
    inspections = await _load_inspections(session, rows)
    await session.commit()

    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(
        [
            "created_at",
            "actor_user_id",
            "actor_name",
            "action",
            "subject",
            "register",
            "site_code",
            "bus_no",
            "object_type",
            "object_id",
            "after",
        ]
    )
    for r in rows:
        out = _audit_out(
            r, actors, entry_by_id=entries, inspection_by_id=inspections
        )
        created = out.created_at
        created_s = (
            created.isoformat() if isinstance(created, datetime) else str(created)
        )
        writer.writerow(
            [
                created_s,
                out.actor_user_id or "",
                out.actor_name or "",
                out.action,
                out.subject or "",
                out.register or "",
                out.site_code or "",
                out.bus_no or "",
                out.object_type,
                out.object_id,
                json.dumps(out.after or {}, default=str),
            ]
        )

    return Response(
        content=buf.getvalue(),
        media_type="text/csv",
        headers={
            "Content-Disposition": 'attachment; filename="enm-audit.csv"',
        },
    )

File: app/lib/models/admin_estate.dart
import 'package:flutter/foundation.dart';

/// Estate-wide Admin Summary — one site's counts for a period.
@immutable
class SiteSummaryMetrics {
  const SiteSummaryMetrics({
    required this.siteCode,
    required this.name,
    required this.isActive,
    required this.operatingCategories,
    this.vehicleCount = 0,
    this.userCount = 0,
    this.workDone = 0,
    this.driverComplaints = 0,
    this.breakdowns = 0,
    this.coolant = 0,
    this.inspections = 0,
    this.openOffRoad = 0,
  });

  final String siteCode;
  final String name;
  final bool isActive;
  final List<String> operatingCategories;
  final int vehicleCount;
  final int userCount;
  final int workDone;
  final int driverComplaints;
  final int breakdowns;
  final int coolant;
  final int inspections;
  final int openOffRoad;

  factory SiteSummaryMetrics.fromJson(Map<String, dynamic> json) {
    return SiteSummaryMetrics(
      siteCode: json['site_code'] as String? ?? '',
      name: json['name'] as String? ?? '',
      isActive: json['is_active'] as bool? ?? true,
      operatingCategories: [
        for (final c in (json['operating_categories'] as List<dynamic>? ?? const []))
          c.toString(),
      ],
      vehicleCount: (json['vehicle_count'] as num?)?.toInt() ?? 0,
      userCount: (json['user_count'] as num?)?.toInt() ?? 0,
      workDone: (json['work_done'] as num?)?.toInt() ?? 0,
      driverComplaints: (json['driver_complaints'] as num?)?.toInt() ?? 0,
      breakdowns: (json['breakdowns'] as num?)?.toInt() ?? 0,
      coolant: (json['coolant'] as num?)?.toInt() ?? 0,
      inspections: (json['inspections'] as num?)?.toInt() ?? 0,
      openOffRoad: (json['open_off_road'] as num?)?.toInt() ?? 0,
    );
  }
}

@immutable
class EstateTotals {
  const EstateTotals({
    this.sites = 0,
    this.activeSites = 0,
    this.vehicles = 0,
    this.users = 0,
    this.workDone = 0,
    this.driverComplaints = 0,
    this.breakdowns = 0,
    this.coolant = 0,
    this.inspections = 0,
    this.openOffRoad = 0,
  });

  final int sites;
  final int activeSites;
  final int vehicles;
  final int users;
  final int workDone;
  final int driverComplaints;
  final int breakdowns;
  final int coolant;
  final int inspections;
  final int openOffRoad;

  factory EstateTotals.fromJson(Map<String, dynamic> json) {
    return EstateTotals(
      sites: (json['sites'] as num?)?.toInt() ?? 0,
      activeSites: (json['active_sites'] as num?)?.toInt() ?? 0,
      vehicles: (json['vehicles'] as num?)?.toInt() ?? 0,
      users: (json['users'] as num?)?.toInt() ?? 0,
      workDone: (json['work_done'] as num?)?.toInt() ?? 0,
      driverComplaints: (json['driver_complaints'] as num?)?.toInt() ?? 0,
      breakdowns: (json['breakdowns'] as num?)?.toInt() ?? 0,
      coolant: (json['coolant'] as num?)?.toInt() ?? 0,
      inspections: (json['inspections'] as num?)?.toInt() ?? 0,
      openOffRoad: (json['open_off_road'] as num?)?.toInt() ?? 0,
    );
  }
}

@immutable
class SegmentSummary {
  const SegmentSummary({
    required this.key,
    required this.label,
    required this.totals,
    required this.sites,
  });

  final String key;
  final String label;
  final EstateTotals totals;
  final List<SiteSummaryMetrics> sites;

  factory SegmentSummary.fromJson(String key, Map<String, dynamic> json) {
    final totalsRaw = json['totals'] as Map<String, dynamic>? ?? const {};
    return SegmentSummary(
      key: key,
      label: json['label'] as String? ?? key,
      totals: EstateTotals(
        sites: (totalsRaw['site_count'] as num?)?.toInt() ?? 0,
        workDone: (totalsRaw['work_done'] as num?)?.toInt() ?? 0,
        driverComplaints: (totalsRaw['driver_complaints'] as num?)?.toInt() ?? 0,
        breakdowns: (totalsRaw['breakdowns'] as num?)?.toInt() ?? 0,
        coolant: (totalsRaw['coolant'] as num?)?.toInt() ?? 0,
        inspections: (totalsRaw['inspections'] as num?)?.toInt() ?? 0,
        openOffRoad: (totalsRaw['open_off_road'] as num?)?.toInt() ?? 0,
      ),
      sites: [
        for (final s in (json['sites'] as List<dynamic>? ?? const []))
          SiteSummaryMetrics.fromJson(s as Map<String, dynamic>),
      ],
    );
  }
}

@immutable
class AdminSummary {
  const AdminSummary({
    required this.dateFrom,
    required this.dateTo,
    required this.estate,
    required this.sites,
    required this.segments,
  });

  final String dateFrom;
  final String dateTo;
  final EstateTotals estate;
  final List<SiteSummaryMetrics> sites;
  final Map<String, SegmentSummary> segments;

  factory AdminSummary.fromJson(Map<String, dynamic> json) {
    final segs = <String, SegmentSummary>{};
    final rawSegs = json['segments'] as Map<String, dynamic>? ?? const {};
    for (final e in rawSegs.entries) {
      segs[e.key] = SegmentSummary.fromJson(
        e.key,
        e.value as Map<String, dynamic>,
      );
    }
    return AdminSummary(
      dateFrom: json['date_from']?.toString() ?? '',
      dateTo: json['date_to']?.toString() ?? '',
      estate: EstateTotals.fromJson(
        json['estate'] as Map<String, dynamic>? ?? const {},
      ),
      sites: [
        for (final s in (json['sites'] as List<dynamic>? ?? const []))
          SiteSummaryMetrics.fromJson(s as Map<String, dynamic>),
      ],
      segments: segs,
    );
  }
}

@immutable
class AuditLogEntry {
  const AuditLogEntry({
    required this.id,
    required this.action,
    required this.objectType,
    required this.objectId,
    required this.createdAt,
    this.actorId,
    this.actorUserId,
    this.actorName,
    this.before,
    this.after,
  });

  final String id;
  final String? actorId;
  final String? actorUserId;
  final String? actorName;
  final String action;
  final String objectType;
  final String objectId;
  final Map<String, dynamic>? before;
  final Map<String, dynamic>? after;
  final String createdAt;

  String get actorLabel {
    if (actorName != null && actorName!.isNotEmpty) {
      final id = actorUserId;
      return id == null || id.isEmpty ? actorName! : '$actorName ($id)';
    }
    return actorUserId ?? actorId ?? '—';
  }

  factory AuditLogEntry.fromJson(Map<String, dynamic> json) {
    Map<String, dynamic>? asMap(Object? v) {
      if (v is Map<String, dynamic>) return v;
      return null;
    }

    return AuditLogEntry(
      id: json['id']?.toString() ?? '',
      actorId: json['actor_id']?.toString(),
      actorUserId: json['actor_user_id']?.toString(),
      actorName: json['actor_name']?.toString(),
      action: json['action']?.toString() ?? '',
      objectType: json['object_type']?.toString() ?? '',
      objectId: json['object_id']?.toString() ?? '',
      before: asMap(json['before']),
      after: asMap(json['after']),
      createdAt: json['created_at']?.toString() ?? '',
    );
  }
}

@immutable
class AuditLogPage {
  const AuditLogPage({
    required this.items,
    required this.page,
    required this.pageSize,
    required this.total,
  });

  final List<AuditLogEntry> items;
  final int page;
  final int pageSize;
  final int total;

  factory AuditLogPage.fromJson(Map<String, dynamic> json) {
    return AuditLogPage(
      items: [
        for (final i in (json['items'] as List<dynamic>? ?? const []))
          AuditLogEntry.fromJson(i as Map<String, dynamic>),
      ],
      page: (json['page'] as num?)?.toInt() ?? 1,
      pageSize: (json['page_size'] as num?)?.toInt() ?? 50,
      total: (json['total'] as num?)?.toInt() ?? 0,
    );
  }
}
